text/0000-as-cell.md

```diff
@@ -0,0 +1,243 @@
+- For each element, access an element a number of elements before or after it.
+
+Todays answer for algorithms like that is to fall back to C-style
+for loops and indexing, which might look like this...
+
+```rust
+
+let v: Vec<i32> = ...;
+
+// example 1
+for i in 0..v.len() {
+    for j in 0..v.len() {
+        v[j] = f(v[i], v[j]);
+    }
+}
+
+// example 2
+for i in n..v.len() {
+    v[i] = g(v[i - n]);
+}
+
+```
+
+...but this reintroduces the bound-checking costs.
+
+The alternative, short of changing the actual algorithms involved, is to use
+internal mutability to enable safe mutations even with overlapping shared views into the data:
+
+```rust
+let v: Vec<Cell<i32>> = ...;
+
+// example 1
+for i in &v {
+    for j in &v {
+        j.set(f(i.get(), j.get()));
+    }
+}
+
+// example 2
+for (i, j) in v[n..].iter().zip(&v) {
+    i.set(g(g.get()));
+}
+
+```
+
+This has the advantages of allowing both bound-check free iteration and
+aliasing references, but comes with restrictions that makes it not generally
+applicable, namely:
+
+- The need to modify the data structure containing the data (Which might not
+  always be possible because it might come from external code).
+- Loss of the ability to directly hand out `&T` and `&mut T` references to the data.
+- Refcounting overhead in case of `RefCell<T>`.
+
+This RFC proposes a way to address these in cases where `Cell<T>`
+would be used by introducing simple conversions functions
+to the standard library that allow the creation of shared borrowed
+`Cell<T>`s from mutably borrowed `T`s.
+
+This in turn allows the original data structure to remain unchanged,
+while allowing to temporary opt-in to the `Cell` API as needed.
+As an example, given `Cell::from_mut_slice(&mut [T]) -> &[Cell<T>]`,
+the previous examples can be written as this:
+
+```rust
+let mut v: Vec<i32> = ...;
+
+// convert the mutable borrow
+let v_slice: &[Cell<T>] = Cell::from_mut_slice(&mut v);
+
+// example 1
+for i in v_slice {
+    for j in v_slice {
+        j.set(f(i.get(), j.get()));
+    }
+}
+
+// example 2
+for (i, j) in v_slice[n..].iter().zip(v_slice) {
+    i.set(g(g.get()));
+}
+
+```
+
+# Detailed design
+[design]: #detailed-design
+
+## Language
+
+In order for this proposal to be safe, __it needs to be guaranteed that
+`T` and `Cell<T>` have the same memory layout__, and that there are no codegen
+issues based on viewing a reference to a `UnsafeCell`-less types as a
+reference to a `UnsafeCell`-containing type.
+
+As far as the author is aware, both should already implicitly
+fall out of the semantic of `Cell` and llvms notion of aliasing:
+
+- `Cell` is safe interior mutability based on memcopying the `T`,
+  and thus does not need additional fields.
+- `&mut T -> &U` is a sub borrow, which prevents access to the original `&mut T`
+  for its duration, thus no aliasing.
+
+However, we might have to add an explicit language guarantee about
+the validity of this cast.
```

Comment:
Sorry, then I don't understand what your initial comment was referring to.

My point with that line in the RFC is that I don't know if the current language spec (aka "what should happen") already validates this specific cast on its own, or if there needs to be a amendment to allow it "officially", rather than just by the implementation of the compiler.